### ISSUE TITLE
Infer Future<void> for await Future.delayed()

### ISSUE THREAD

**Issue by nex3** [dart-lang/language#1398]
Currently, if you write `await Future.delayed(Duration(milliseconds: 100))`, the analyzer produces an `inference_failure_on_instance_creation` hint saying that the type argument can't be inferred. This is annoying, since the type argument doesn't matter at all for a bare await statement. Would it be possible for type inference to instead infer this as `Future<void>`?

**Comment by mateusfccp** [comment]
It is not possible for the analyzer to infer a type that is not true. `await Future.delayed(...)` is not `Future<void>`. It depends on the type parameter `T` that is used as the generic for the return type of the `computation` parameter. As stated in the documentation, `computation` is treated as `() => null` when omitted.

The analyzer can't infer the type of `() => null`, this is why you get the `inference_failure_on_instance_creation` hint. However, you can quickly fix it by providing the type parameter.

```dart
await Future<void>.delayed(const Duration(milliseconds: 100))
```

**Comment by rrousselGit** [comment]
It'd be feasible if `Future.delayed` was a static method instead of a factory:

```dart
class Future<T> {
  static Future<void> delayed(Duration duration) => ...
}
```

**Comment by srawlins** [comment]
@mateusfccp 

> The analyzer can't infer the type of () => null

The default closure is actually `null`, because `() => null` is not a compile-time constant. `() => null` is used later.

@rrousselGit 

> It'd be feasible if Future.delayed was a static method instead of a factory

This would work but would split the use into one with the `computation` parameter, and one without. We've had to do this for other functions, like `runZoned`.

**Comment by mateusfccp** [comment]
@srawlins 

> The default closure is actually `null`, because `() => null` is not a compile-time constant. `() => null` is used later.

Yes, of course. We don't have constant function literals (#1048). The way I wrote implied that the analyzer tried to infer the type of `() => null`, which is wrong. Sorry!

Does the analyzer always infer `dynamic` for type parameters when `null` is passed?

**Comment by lrhn** [comment]
We could, perhaps, make the context type of the expression of an expression statement be `void`.
Then `await Future.delayed(Duration.zero)` would have context type `void`, and `Future.delayed(...)` would have context type `FutureOr<void>`, which *could* mean that the future has type `Future<void>`.

Currently we don't give it any context type, which is why it has to guess from the parameter.
Example (tested on nullsafety.dartpad.dev):
```dart
import "dart:async";
T id<T>(T value) => value;

main() async {
  /* void x = */ await (contextType(null)..checkStaticType<Exactly<FutureOr<void>>>());
}

extension <T> on T {
  void checkStaticType<R extends Exactly<T>>() {
    print(T);
  }
}
typedef Exactly<T> = T Function(T);
T contextType<T>(Object? o) => o as T;
```
This code prints `Object?` as the static type of `contextType(null)`, which means no context type.
If I uncomment the `void x =`, the context type of the awaited expression ` becomes `void` (probably normalized from `FutureOr<void>` because I'm allowed to call methods on it, which I wouldn't if it was pure `void`).
